Add ListUsers to the services.IdentityService

rosstimothy opened the pull request:
Listing users is only implemented in the cache and backend implementation of the interface. All clients and servers return not implemented and will be updated in a future PR. Because the list operation is so unique for users due to user resources, and user secrets being stored at the same prefix I wanted it to be reviewed on its own.

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +73,83 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+	rangeEnd := backend.RangeEnd(backend.ExactKey(webPrefix, usersPrefix))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > 1000 {
+		pageSize = 1000
+	}
```
```suggestion
	if pageSize <= 0 || pageSize > apidefaults.DefaultChunkSize {
		pageSize = apidefaults.DefaultChunkSize
	}
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +73,83 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+		// If the number of users retrieved is more than the requested page size
+		// then return the number of users requested and the token for the user
+		// at page size + 1 in the slice.
+		if diff := len(out) - pageSize; diff > 0 {
+			nextKey := backend.GetPaginationKey(out[len(out)-diff])
+			// Truncate the last item that was used to determine next row existence.
+			out = out[:pageSize]
+			return out, nextKey, nil
+		}
```
```suggestion
		if len(out) > pageSize {
			nextKey := backend.GetPaginationKey(out[pageSize])
			// Truncate the last item that was used to determine next row existence.
			clear(out[pageSize:])
			return out[:pageSize], nextKey, nil
		}
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +73,83 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+		}
+
+		if withSecrets {
+			collected, _, err := collectUserItems(result.Items)
+			if err != nil {
+				return nil, "", trace.Wrap(err)
+			}
+
+			for uname, uitems := range collected {
+				user, err := userFromUserItems(uname, uitems)
+				if err != nil {
+					return nil, "", trace.Wrap(err)
+				}
+				out = append(out, user)
+			}
```
Why are we guaranteed to get the full user data from a `GetRange` that might end halfway through the items belonging to the user?

espadolini commented on lib/cache/cache.go:
```diff
@@ -2015,6 +2015,24 @@ func (c *Cache) GetUsers(ctx context.Context, withSecrets bool) ([]types.User, e
 	return rg.reader.GetUsers(ctx, withSecrets)
 }
 
+// ListUsers returns the next page of users.
```
```suggestion
// ListUsers returns a page of users.
```

espadolini commented on lib/services/identity.go:
```diff
@@ -60,6 +60,8 @@ type UsersService interface {
 	DeleteUser(ctx context.Context, user string) error
 	// GetUsers returns a list of users registered with the local auth server
 	GetUsers(ctx context.Context, withSecrets bool) ([]types.User, error)
+	// ListUsers returns the next page of users.
```
```suggestion
	// ListUsers returns a page of users.
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+		// start key matches the last item in the results then this page has already been
+		// retried once and means there are no more user entries to search so return the
+		// partial page of users.
+		nextStart := []byte(backend.NextPaginationKey(paginationKeyResource{name: string(result.Items[len(result.Items)-1].Key)}))
```
Isn't this literally just `backend.RangeEnd(result.Items[len(result.Items)-1].Key)` (in lieu of `backend.nextKey`)?

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+	return f.name
+}
+
+func (s *IdentityService) listUsersWitSecrets(ctx context.Context, pageSize int, nextToken string) ([]types.User, string, error) {
```
```suggestion
func (s *IdentityService) listUsersWithSecrets(ctx context.Context, pageSize int, nextToken string) ([]types.User, string, error) {
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+// ListUsers returns the next page of users.
+func (s *IdentityService) ListUsers(ctx context.Context, pageSize int, nextToken string, withSecrets bool) ([]types.User, string, error) {
+	if withSecrets {
+		users, next, err := s.listUsersWitSecrets(ctx, pageSize, nextToken)
```
```suggestion
		users, next, err := s.listUsersWithSecrets(ctx, pageSize, nextToken)
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+		}
+
+		for _, item := range result.Items {
+			if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
```
This was in the original `GetUsers` as well, but... are we sure that there's no way to get a MFA device in there that has an ID of `params`?

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+			}
+
+			// Prevent adding duplicate resources
+			if len(out) != 0 && out[len(out)-1].GetName() == resource.GetName() {
+				continue
+			}
```
When can this happen?

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+			out = append(out, resource)
+		}
+
+		// If the number of users retrieved is more than the requested page size
+		// then return the number of users requested and the token for the user
+		// at page size + 1 in the slice.
```
Aren't we risking an information leak if we return the name of the next user rather than the key after the last returned user? Can anyone call `ListUsers` without having full read on all users anyway?

espadolini reviewed (commented):
I feel like this would read way better if we used `backend.StreamRange`; the one without secrets can just be a `FilterMap` and a `Take`, and the one with secrets probably needs a for loop but it can just accumulate stuff until the prefix changes or the listing ends, then collect the items from the user together.

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +75,125 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+// the next user in the list while still allowing listing to operate
+// without missing any users.
+func nextUserToken(user types.User) string {
+	return string(backend.RangeEnd(backend.ExactKey(user.GetName())))[1:]
```
This might be clearer.

```suggestion
	return string(backend.RangeEnd(user.GetName() + backend.Separator))
```

espadolini commented on lib/services/local/users.go:
```diff
@@ -73,6 +75,125 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+	// If there are no more users its possible that the previous user being processed
+	// was never added to out because there was no additional user to transition to.
+	// Add the user from the collected items and return the full output.
+	if len(out) > 0 && out[len(out)-1].GetName() != prevUser {
```
Won't this fail to return anything if there's only a single user? A simpler check would be to see if `items` has anything:

```suggestion
	if items.complete() || items.hasLocalAuthItems() {
```

Maybe we could add an `isEmpty` method to `userItems` and use it here.

rosstimothy committed:
Add ListUsers to the services.IdentityService

Listing users is different from most existing list operations
because multiple resource types are stored under the users prefix.
Care was taken to ensure that these extra resources(passwords,
mfa devices, etc) do not cause lising to omit any users.
